### PR TITLE
Ipv6 private extensions

### PR DESCRIPTION
**Pull request by ssahani**

### CODE REVIEW

**Comment by keszybz** [comment]
Looks good, apart from the description. What this option does is enable temporary addresses. I think it is more useful to say that, instead of talking about "privacy extensions", which is not really clear.

RFC 3041 was obsoleted by RFC 4941. Shouldn't the link be updated?

In the man page, say something like

```
Configures use of stateless temporary addresses that change over time
(see <ulink url="https://tools.ietf.org/html/rfc4941">RFC 4941</ulink>, Privacy Extensions for Stateless Address Autoconfiguration in IPv6). Takes one of three values. ...
```

File: src/network/networkd.h [hunk 1]
```diff
@@ -90,6 +90,14 @@ typedef enum DCHPClientIdentifier {
         _DHCP_CLIENT_ID_INVALID = -1,
 } DCHPClientIdentifier;
 
+typedef enum Ipv6PrivacyExtensions {
+        IPV6_PRIVACY_EXTENSIONS_DISABLE,
+        IPV6_PRIVACY_EXTENSIONS_PREFER_PUBLIC,
+        IPV6_PRIVACY_EXTENSIONS_PREFER_TEMPORARY,
+        _IPV6_PRIVACY_EXTENSIONS_MAX,
+        _IPV6_PRIVACY_EXTENSIONS_INVALID = -1,
+} Ipv6PrivacyExtensions;
```
Comment (zonque) [inline src/network/networkd.h]: That might be nit-picking, but I think it should consistently be called "IPv6", not "Ipv6".

File: src/network/networkd-network.c [hunk 2]
```diff
@@ -751,3 +753,15 @@ int config_parse_address_family_boolean_with_kernel(
...
+        [IPV6_PRIVACY_EXTENSIONS_DISABLE] = "disable",
+        [IPV6_PRIVACY_EXTENSIONS_PREFER_PUBLIC] = "prefer-public",
+        [IPV6_PRIVACY_EXTENSIONS_PREFER_TEMPORARY] = "prefer-temporary",
+};
+
+DEFINE_STRING_TABLE_LOOKUP(ipv6_privacy_extensions, Ipv6PrivacyExtensions);
+DEFINE_CONFIG_PARSE_ENUM(config_parse_ipv6_privacy_extensions,
+                         ipv6_privacy_extensions,
+                         Ipv6PrivacyExtensions,
+                         "Failed to parse Ipv6 Privacy Extensions");
```
Comment (zonque) [inline src/network/networkd-network.c]: "IPv6", not "Ipv6"

**Comment by ssahani** [comment]
Thanks @keszybz for the manual and  @zonque  for review.

updated

File: src/network/networkd.h [hunk 3]
```diff
@@ -145,6 +153,8 @@ struct Network {
         AddressFamilyBoolean ip_forward;
         bool ip_masquerade;
 
+        IPV6PrivacyExtensions ipv6_privacy_extensions;
```
Comment (zonque) [inline src/network/networkd.h]: No, the "v" is lower case, "IP" is uppercase :)
https://en.wikipedia.org/wiki/IPv6

Comment (ssahani) [inline src/network/networkd.h]: Yes thanks !

**Comment by ssahani** [comment]
Updated now.

**Comment by zonque** [comment]
Looks good to me now. @teg, any final opinion? We might think about if should be called `disabled` rather than `disable` or something?

**Comment by poettering** [comment]
Patch set looks pretty good to me. However, I wonder if we should really expose this as enum. Pretty much all examples I can find on the internet, only set the value to either 0 or 2, never to 1. i.e. the prefer-public mode is almost never used. I also fail to see what the point of it is: i mean if we do the whole exercise why then ever use the prefer-public setting at all?

Hence I think that the option parser should really consider this to be a boolean with one extra value. I.e. "yes", "1", "on", should mean that the setting is set to "2", and "no", 0", "off" should mean that the setting is set to "0". This should be attempted to be parsed via parse_boolean(). If parse_boolean() fails then we should check that against the "prefer-public" string before considering this a parse failure. Or with other words: internally is should be an enum like you already have it in place in your patch, but the parser should not be a normal enum parser, but try to parse the setting as boolean and only if that fails make one extra check for the third enum value.

See config_parse_llmnr() for an example how to do this. The LLMNR= setting also is a boolean with an extra enum value.

Hope that makes sense?

**Comment by zonque** [comment]
Yeah, I agree. That would also explain users that "prefer-temporary" is the one that actually makes sense, and the "prefer-public" is supported but does not actually enforcing privacy. That should be reflected in the docs as well.

**Comment by ssahani** [comment]
@poettering thanks for review. Makes sense . I have updated the patch.

**Comment by zonque** [comment]
Looks good to me. Merged, thanks!

**Comment by poettering** [comment]
Hmm, the patch isn't right. It now maps the boolean true to "prefer-public". It should map to "prefer-temporary". Also, the enum values should really match the setting names. There are also whitespace errors. And the setting should always be written, not only when set != "no".

I'lll fix that.

**Comment by poettering** [comment]
Also, link_ipv6_privacy_extensions_enabled() returns a bool right now, but the enum is sometimes returned by it, which is at least confusing, but probably wrong.

**Comment by ssahani** [comment]
I think link_ipv6_privacy_extensions_enabled values are -1 0 1 2 
So if we return the enum disabled as false (0) and 1 and 2 interpreted as true.  which is prefer public and prefer temporary

**Comment by poettering** [comment]
I fixed a number of issues now in #496. Please review.

**Comment by ssahani** [comment]
Looks great thanks !

**Pull request merged**